module/VuFind/src/VuFind/View/Helper/Root/MakeTag.php

```diff
@@ -0,0 +1,157 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  View_Helpers
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+namespace VuFind\View\Helper\Root;
+
+/**
+ * Make tag view helper
+ *
+ * @category VuFind
+ * @package  View_Helpers
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class MakeTag extends \Laminas\View\Helper\AbstractHelper
+{
+    /**
+     * List of all void tags (tags that access no innerHTML)
+     *
+     * Source: https://html.spec.whatwg.org/multipage/syntax.html#void-elements
+     *
+     * @var string[]
+     */
+    protected $voidElements = [
+        'area',
+        'base',
+        'br',
+        'col',
+        'embed',
+        'hr',
+        'img',
+        'input',
+        'link',
+        'meta',
+        'param',
+        'source',
+        'track',
+        'wbr'
+    ];
+
+    /**
+     * Render an HTML tag
+     *
+     * A string passed into $attrs will be treated like a class.
+     * These two are equivalent:
+     * > MakeTag('div', 'Success!', 'alert alert-success')
+     * > MakeTag('div', 'Success!', ['class => 'alert alert-success'])
+     *
+     * Additional options
+     * - escapeContent: Default true, set to false to skip escaping (like for HTML).
+     *
+     * @param string       $tagName  Element tag name
+     * @param string       $contents Element contents (must be properly-formed HTML)
+     * @param string|array $attrs    Tag attributes (associative array or class name)
+     * @param array        $options  Additional options
+     *
+     * @return string HTML for an anchor tag
+     */
+    public function __invoke(
+        string $tagName,
+        string $contents,
+        $attrs = [],
+        $options = []
+    ) {
+        // $attrs not an object, interpret as class name
+        if (!is_array($attrs)) {
+            $attrs = !empty($attrs) ? ['class' => $attrs] : [];
+        }
+
+        // Compile attributes
+        return $this->compileTag($tagName, $contents, $attrs, $options);
+    }
+
+    /**
+     * Verify HTML tag matches HTML spec:
+     * - starts with letter
+     * - contains only letters and hyphens
+     *
+     * @param string $tagName  Element tag name
+     *
+     * @return void
+     * @throws InvalidArgumentException
+     */
+    protected function verifyTagName(string $tagName)
+    {
+        $validTagPattern = '/^[a-zA-Z][a-zA-Z\-]*$/';
```

Comment:
Is this the right validation pattern? From the HTML5 spec: "HTML elements all have names that only use characters in the range U+0030 DIGIT ZERO (0) to U+0039 DIGIT NINE (9), U+0061 LATIN SMALL LETTER A to U+007A LATIN SMALL LETTER Z, and U+0041 LATIN CAPITAL LETTER A to U+005A LATIN CAPITAL LETTER Z."

Reply:
I was looking at [this spec](https://html.spec.whatwg.org/multipage/parsing.html#tag-name-state) which doesn't mention numbers and [this spec](https://html.spec.whatwg.org/multipage/parsing.html#tag-open-state) which requires a letter as the first character. So my understanding was:
- start with a letter
- can include hyphens (i.e. custom tags which require one)
- can be capital or lowercase or mixed (although they are all apparently parses as lowercase ultimately)

Reply:
@crhallberg, I think I interpret 13.2.5.8 differently than you did. Note that the last clause is for "Anything else," which gets appended to the tag name. This actually seems to suggest that the range of legal tag name characters is very broad. As you say, the name must start with an alphabetical character, but it seems that the following characters can be anything except whitespace, forward slash, or greater than sign. I think at very least this suggests we should add numbers (and perhaps also underscores) into the valid tag pattern.

Reply:
I see your point! This contradicts what I understand about custom elements and I realized there is a further spec about what is considered valid (outside what the parser spec will accept).

- [Valid tags inside `<body>`](https://html.spec.whatwg.org/multipage/dom.html#flow-content)
- [Custom element names](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name)

Reply:
It's amazing how complicated something so apparently simple can become! If nothing else, I think this is an indicator that we should define the valid tag pattern regex as a class property so it's easy to customize/override.